tachyon/crypto/sumcheck/sumcheck_prover.h

```diff
@@ -0,0 +1,180 @@
+  // [XZZPS19](https://eprint.iacr.org/2019/317.pdf#subsection.3.2).
+  //
+  // Referencing https://people.cs.georgetown.edu/jthaler/sumcheck.pdf
+  // For v = the number of variables...
+  // For the first round, the Prover creates and sends:
+  //   g₁(X₁) := ∑ g(X₁,x₂,...,xᵥ), (x₂,...,xᵥ)∈{0,1}ᵛ⁻¹
+  // For the consequent round i, the Prover creates and sends:
+  //   gᵢ(Xᵢ) = ∑ g(r₁,...,rᵢ₋₁,Xᵢ,xᵢ₊₁,...,xᵥ), (xᵢ₊₁ ,...,xv)∈{0,1}ᵛ⁻¹
+  // More specifically, |Prover(VerifierMsg)| fixes the front variables with
+  // r₁,...,rᵢ₋₁, and |Prover()| creates the total gᵢ(Xᵢ).
+  SumcheckProverMsg<F, MaxDegree> Round() {
+    // Max number of prover rounds should be |num_variables_|.
+    CHECK_LE(++round_, num_variables_);
+
+    // Generate sum
+    // clang-format off
+    // g(x₁, x₂) = 1(1 - x₁)(1 - x₂) + 2x₁(1 - x₂) + 3(1 - x₁)x₂ + 4x₁x₂
+    //
+    // Creating the univariate polynomial:
+    // g₁(X₁) = ∑ 1(1 - X₁)(1 - x₂) + 2X₁(1 - x₂) + 3(1 - X₁)x₂ + 4X₁x₂, x₂∈{0,1}
+    //        = 1(1 - X₁) + 2X₁ + 3(1 - X₁) + 4X₁
+    //        = start₀ + step₀X₁ + start₁ + step₁X₁
+    //         (where start₀ = 1, step₀ = 1 = 2 - 1, start₁ = 3 and step₁ = 1 = 4 - 3)
+    //        = start₀ + start₁ + (step₀ + step₁)X₁
+    //
+    // g₁(0) = start₀ + start₁
+    // g₁(1) = start₀ + start₁ + step₀ + step₁
+    // g₁(2) = start₀ + start₁ + 2step₀ + 2step₁
+    // g₁(3) = start₀ + start₁ + 3step₀ + 3step₁
+    //         (where product_sum₀ = start₀ + start₁ and product_sum₁ = product_sum₀ + step₀ + step₁)
+    // clang-format on
+#if defined(TACHYON_HAS_OPENMP)
+    size_t thread_nums = static_cast<size_t>(omp_get_max_threads());
+#else
+    size_t thread_nums = 1;
+#endif
+    size_t size = size_t{1} << (num_variables_ - round_);
+    thread_nums = (thread_nums * kParallelFactor) <= size ? thread_nums : 1;
+
+    size_t chunk_size = (size + thread_nums - 1) / thread_nums;
+    size_t num_chunks = (size + chunk_size - 1) / chunk_size;
+
+    std::vector<std::vector<F>> products_sums(
+        num_chunks, std::vector<F>(max_evaluations_ + 1, F::Zero()));
+
+    OPENMP_PARALLEL_FOR(size_t i = 0; i < num_chunks; ++i) {
+      size_t begin = i * chunk_size;
+      size_t len = (i == num_chunks - 1) ? size - begin : chunk_size;
+      std::vector<F> products(max_evaluations_ + 1, F::Zero());
+      for (size_t j = begin; j < begin + len; ++j) {
+        for (const Term& term : terms_) {
+          std::fill(products.begin(), products.end(), term.coefficient);
+          CalculateProducts(j, products, term);
+          for (size_t k = 0; k < max_evaluations_ + 1; ++k) {
+            products_sums[i][k] += products[k];
+          }
+        }
+      }
+    }
+    for (size_t i = 1; i < num_chunks; ++i) {
+      for (size_t j = 0; j < max_evaluations_ + 1; ++j) {
+        products_sums[0][j] += products_sums[i][j];
+      }
+    }
+    return {
+        math::UnivariateEvaluations<F, MaxDegree>(std::move(products_sums[0]))};
```

Comment:
This is strangely wrapped

```suggestion
    return {{std::move(products_sums[0]))}};
```